File: backend/app/schemas/supervisor_feedback.py
from typing import Optional, TYPE_CHECKING
from uuid import UUID
from pydantic import BaseModel, Field, field_validator
from datetime import datetime
from .common import SubmissionStatus, RatingCode, PaginatedResponse
from .supervisor_review import SupervisorAction

if TYPE_CHECKING:
    from .self_assessment import SelfAssessment
    from .evaluation import EvaluationPeriod
    from .user import UserProfileOption


class SupervisorFeedbackBase(BaseModel):
    supervisor_rating_code: Optional[RatingCode] = Field(
        None,
        alias="supervisorRatingCode",
        description="Supervisor's letter grade: SS, S, A, B, C, D (6-level input scale)"
    )
    supervisor_comment: Optional[str] = Field(
        None,
        alias="supervisorComment",
        max_length=5000,
        description="Supervisor's feedback comment"
    )


class SupervisorFeedbackCreate(SupervisorFeedbackBase):
    """Request schema for creating supervisor feedback."""
    self_assessment_id: UUID = Field(..., alias="selfAssessmentId")
    period_id: UUID = Field(..., alias="periodId")
    rating_data: Optional[dict] = Field(
        None,
        alias="ratingData",
        description="Granular per-action ratings for コンピテンシー goals (JSONB). NULL for 業績目標."
    )
    action: SupervisorAction = Field(
        default=SupervisorAction.PENDING,
        description="Decision: PENDING or APPROVED"
    )
    status: SubmissionStatus = Field(
        default=SubmissionStatus.DRAFT,
        description="Workflow status: incomplete, draft, or submitted"
    )

    model_config = {"populate_by_name": True}


class SupervisorFeedbackUpdate(BaseModel):
    """Request schema for updating supervisor feedback."""
    supervisor_rating_code: Optional[RatingCode] = Field(
        None,
        alias="supervisorRatingCode",
        description="Supervisor's letter grade: SS, S, A, B, C, D (6-level input scale)"
    )
    supervisor_comment: Optional[str] = Field(
        None,
        alias="supervisorComment",
        max_length=5000,
        description="Supervisor's feedback comment"
    )
    rating_data: Optional[dict] = Field(
        None,
        alias="ratingData",
        description="Granular per-action ratings for コンピテンシー goals (JSONB). NULL for 業績目標."
    )

    model_config = {"populate_by_name": True}


class SupervisorFeedbackSubmit(BaseModel):
    """Request schema for submitting supervisor feedback."""
    action: SupervisorAction = Field(
        ...,
        description="Decision: PENDING or APPROVED (REJECTED is not supported)"
    )
    supervisor_rating_code: Optional[RatingCode] = Field(
        None,
        alias="supervisorRatingCode",
        description="Supervisor's letter grade (optional)"
    )
    supervisor_comment: Optional[str] = Field(
        None,
        alias="supervisorComment",
        max_length=5000,
        description="Supervisor's feedback comment (optional)"
    )
    rating_data: Optional[dict] = Field(
        None,
        alias="ratingData",
        description="Granular per-action ratings for コンピテンシー goals (JSONB). NULL for 業績目標."
    )

    @field_validator("action")
    @classmethod
    def validate_action(cls, value: SupervisorAction) -> SupervisorAction:
        if value not in (SupervisorAction.PENDING, SupervisorAction.APPROVED):
            raise ValueError("action must be PENDING or APPROVED")
        return value

    model_config = {"populate_by_name": True}


class SupervisorFeedbackInDB(SupervisorFeedbackBase):
    id: UUID
    self_assessment_id: UUID
    period_id: UUID
    supervisor_id: UUID
    subordinate_id: Optional[UUID] = None
    supervisor_rating: Optional[float] = Field(
        None,
        alias="supervisorRating",
        ge=0,
        le=7,
        description="Numeric rating (0.0-7.0), auto-calculated"
    )
    rating_data: Optional[dict] = Field(
        None,
        alias="ratingData",
        description="Granular per-action ratings for コンピテンシー goals (JSONB). NULL for 業績目標."
    )
    action: SupervisorAction = Field(
        default=SupervisorAction.PENDING,
        description="Decision: PENDING or APPROVED"
    )
    status: SubmissionStatus = Field(
        default=SubmissionStatus.INCOMPLETE,
        description="Workflow status: incomplete, draft, or submitted"
    )
    submitted_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True, "populate_by_name": True}


class SupervisorFeedback(SupervisorFeedbackInDB):
    """
    Basic supervisor feedback schema for API responses (list views, simple references).
    """
    self_assessment_id: UUID = Field(..., alias="selfAssessmentId")
    period_id: UUID = Field(..., alias="periodId")
    supervisor_id: UUID = Field(..., alias="supervisorId")
    subordinate_id: Optional[UUID] = Field(None, alias="subordinateId")
    submitted_at: Optional[datetime] = Field(None, alias="submittedAt")
    reviewed_at: Optional[datetime] = Field(None, alias="reviewedAt")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = {"from_attributes": True, "populate_by_name": True}


class SupervisorFeedbackDetail(SupervisorFeedbackInDB):
    """
    Detailed supervisor feedback schema for single item views.
    """
    self_assessment_id: UUID = Field(..., alias="selfAssessmentId")
    period_id: UUID = Field(..., alias="periodId")
    supervisor_id: UUID = Field(..., alias="supervisorId")
    subordinate_id: Optional[UUID] = Field(None, alias="subordinateId")
    submitted_at: Optional[datetime] = Field(None, alias="submittedAt")
    reviewed_at: Optional[datetime] = Field(None, alias="reviewedAt")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    # Related self-assessment information
    self_assessment: Optional['SelfAssessment'] = Field(None, alias="selfAssessment")

    # Related evaluation period information
    evaluation_period: Optional['EvaluationPeriod'] = Field(
        None,
        alias="evaluationPeriod",
    )

    # User information
    subordinate_info: Optional['UserProfileOption'] = Field(None, alias="subordinateInfo")
    supervisor_info: Optional['UserProfileOption'] = Field(None, alias="supervisorInfo")

    # Feedback state information
    is_editable: bool = Field(True, alias="isEditable")
    is_overdue: bool = Field(False, alias="isOverdue")
    days_until_deadline: Optional[int] = Field(None, alias="daysUntilDeadline")

    # Assessment context
    goal_category: Optional[str] = Field(None, alias="goalCategory")
    goal_title: Optional[str] = Field(None, alias="goalTitle")

    model_config = {"from_attributes": True, "populate_by_name": True}


class SupervisorFeedbackList(PaginatedResponse[SupervisorFeedback]):
    """Schema for paginated supervisor feedback list responses"""
    pass


# ========================================
# FORWARD REFERENCES UPDATE
# ========================================

def rebuild_models():
    """Rebuild models with forward references when all schemas are loaded."""
    try:
        from .self_assessment import SelfAssessment
        from .evaluation import EvaluationPeriod
        from .user import UserProfileOption

        _ = SelfAssessment, EvaluationPeriod, UserProfileOption

        SupervisorFeedbackDetail.model_rebuild()
    except ImportError:
        pass
    except Exception as e:
        print(f"Warning: Could not rebuild forward references in supervisor_feedback schemas: {e}")

rebuild_models()

File: frontend/src/api/types/supervisor-feedback.ts
import type {
  UUID,
  RatingCode,
  SupervisorFeedbackAction,
  SupervisorFeedbackStatus,
  PaginatedResponse,
} from './common';
import type { SelfAssessmentWithGoal } from './self-assessment';
import type { EvaluationPeriod } from './evaluation-period';
import type { UserProfileOption } from './user';

/**
 * Supervisor Feedback type definitions
 * These types match the backend Pydantic schemas for SupervisorFeedback-related operations
 * @see .kiro/specs/self-assessment/frontend-typescript-types.md
 */

/**
 * Base supervisor feedback fields (editable by supervisor)
 */
export interface SupervisorFeedbackBase {
  /** Supervisor's letter grade: SS, S, A+, A, A-, B, C, D */
  supervisorRatingCode?: RatingCode | null;
  /** Supervisor's feedback comment */
  supervisorComment?: string | null;
}

/**
 * Supervisor feedback entity (API response)
 * @see .kiro/specs/self-assessment/api-contract.md Section 5.1
 */
export interface SupervisorFeedback extends SupervisorFeedbackBase {
  /** Unique identifier */
  id: UUID;
  /** Reference to the self-assessment */
  selfAssessmentId: UUID;
  /** Reference to the evaluation period */
  periodId: UUID;
  /** Supervisor who created this feedback */
  supervisorId: UUID;
  /** Subordinate who created the self-assessment */
  subordinateId: UUID;
  /** Numeric rating (0.0-7.0), auto-calculated */
  supervisorRating?: number;
  /** Decision: PENDING or APPROVED */
  action: SupervisorFeedbackAction;
  /** Workflow status: incomplete, draft, or submitted */
  status: SupervisorFeedbackStatus;
  /** Timestamp when feedback was submitted */
  submittedAt?: string;
  /** Timestamp when review was completed */
  reviewedAt?: string;
  /** Record creation timestamp */
  createdAt: string;
  /** Last modification timestamp */
  updatedAt: string;
}

/**
 * Detailed supervisor feedback with additional context
 * @see .kiro/specs/self-assessment/api-contract.md Section 5.3
 *
 * NOTE: Goal title/description should be extracted from selfAssessment.goal
 * using getGoalTitle() and getGoalDescription() helpers.
 * Period name should be extracted from evaluationPeriod.name.
 */
export interface SupervisorFeedbackDetail extends SupervisorFeedback {
  /** The self-assessment this feedback is for (includes goal) */
  selfAssessment?: SelfAssessmentWithGoal;
  /** The evaluation period this feedback belongs to */
  evaluationPeriod?: EvaluationPeriod;
  /** Whether this feedback can still be edited */
  isEditable: boolean;
  /** Whether this feedback is past the deadline */
  isOverdue: boolean;
  /** Days remaining until feedback deadline */
  daysUntilDeadline?: number;
  /** Subordinate who created the self-assessment */
  subordinate?: UserProfileOption;
  /** Supervisor providing the feedback */
  supervisor?: UserProfileOption;
}

/**
 * Request body for creating supervisor feedback
 * @see .kiro/specs/self-assessment/api-contract.md Section 5.2
 */
export interface SupervisorFeedbackCreate extends SupervisorFeedbackBase {
  /** Reference to the self-assessment */
  selfAssessmentId: UUID;
  /** Reference to the evaluation period */
  periodId: UUID;
  /** Initial action (usually PENDING) */
  action: SupervisorFeedbackAction;
  /** Initial status */
  status: SupervisorFeedbackStatus;
}

/**
 * Request body for updating supervisor feedback
 * @see .kiro/specs/self-assessment/api-contract.md Section 5.4
 */
export interface SupervisorFeedbackUpdate {
  /** Supervisor's letter grade */
  supervisorRatingCode?: RatingCode | null;
  /** Supervisor's feedback comment */
  supervisorComment?: string | null;
  /** Per-action ratings for competency goals (JSONB) */
  ratingData?: Record<string, Record<string, RatingCode>>;
}

/**
 * Request body for submitting supervisor feedback (approve)
 * @see .kiro/specs/self-assessment/api-contract.md Section 5.5
 */
export interface SupervisorFeedbackSubmit {
  /** Decision: PENDING or APPROVED */
  action: 'PENDING' | 'APPROVED';
  /** Optional rating code */
  supervisorRatingCode?: RatingCode | null;
  /** Optional comment */
  supervisorComment?: string | null;
  /** Per-action ratings for competency goals (JSONB) */
  ratingData?: Record<string, Record<string, RatingCode>>;
}

/**
 * Paginated list of supervisor feedbacks
 */
export type SupervisorFeedbackList = PaginatedResponse<SupervisorFeedback>;

/**
 * Query parameters for fetching supervisor feedbacks
 */
export interface SupervisorFeedbackQueryParams {
  periodId?: UUID;
  supervisorId?: UUID;
  subordinateId?: UUID;
  selfOnly?: boolean;
  status?: SupervisorFeedbackStatus;
  action?: SupervisorFeedbackAction;
  page?: number;
  limit?: number;
}
